### PR TITLE
wpa_supplicant: add wpa_supplicant

### PR DESCRIPTION
**Pull request by ziemleszcz**
JIRA: DTR-8

### CODE REVIEW

File: wpa_supplicant/patches/04-bswap.patch [hunk 1]
```diff
@@ -0,0 +1,15 @@
+diff -ruN a/src/utils/common.h b/src/utils/common.h
+--- a/src/utils/common.h	2021-05-31 16:31:59.653375453 +0200
++++ b/src/utils/common.h	2021-05-16 11:47:19.467296144 +0200
+@@ -53,6 +53,11 @@
+ }
+ #endif /* __APPLE__ */
+ 
++#ifdef __phoenix__
```
Comment (nalajcie) [inline wpa_supplicant/patches/04-bswap.patch:8]: the `phoenix` define should be provided by the compiler, so we should use it, see:

https://github.com/gcc-mirror/gcc/blob/9faaa091e53db9cb4aee137bb4a6a9c6a18c7af5/gcc/config/phoenix.h#L23

File: wpa_supplicant/build.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+# Make
+#
+mkdir -p "$PREFIX_WPA_SUPPLICANT_INSTALL"
+
+pushd "${PREFIX_WPA_SUPPLICANT_SRC}/wpa_supplicant"
+cp "$PREFIX_WPA_SUPPLICANT/.config" .config
+make install EXTRA_CFLAGS="-D__phoenix__" DESTDIR="$PREFIX_WPA_SUPPLICANT_INSTALL" LIBDIR="/lib" INCDIR="/include" BINDIR="/sbin"
+popd
+
+mkdir -p "$PREFIX_PROG" && cp -Ra "${PREFIX_WPA_SUPPLICANT_INSTALL}/sbin/wpa_cli" "${PREFIX_PROG}/wpa_cli"
```
Comment (nalajcie) [inline wpa_supplicant/build.sh:49]: no need for `mkdir`, they are created by root `build.sh` 

`cp -a`, `-R` is redundant

File: wpa_supplicant/.config [hunk 3]
```diff
@@ -0,0 +1,615 @@
+# Example wpa_supplicant build time configuration
```
Comment (nalajcie) [inline wpa_supplicant/.config:1]: maybe it would be better to store it not as a dot-file (just as a `config`)?

File: wpa_supplicant/build.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    fi
+done
+
+#
+# Make
+#
+mkdir -p "$PREFIX_WPA_SUPPLICANT_INSTALL"
+
+pushd "${PREFIX_WPA_SUPPLICANT_SRC}/wpa_supplicant"
+cp "$PREFIX_WPA_SUPPLICANT/.config" .config
```
Comment (nalajcie) [inline wpa_supplicant/build.sh:45]: `cp -a`, don't update timestamp if original file hasn't changed

File: build.sh [hunk 5]
```diff
@@ -57,4 +57,6 @@ fi
 
 [ "${PORTS_MICROPYTHON}" = "y" ] && ./phoenix-rtos-ports/micropython/build.sh
 
+[ "${PORTS_WPA_SUPPLICANT}" = "y" ] && ./phoenix-rtos-ports/wpa_supplicant/build.sh
```
Comment (nalajcie) [inline build.sh:60]: For me compilation fails with:

```
  CC  ../src/crypto/tls_openssl.c
../src/l2_packet/l2_packet_linux.c:11:10: fatal error: netpacket/packet.h: No such file or directory
   11 | #include <netpacket/packet.h>
```

do I need some other changes for this to work?

File: wpa_supplicant/build.sh [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+make install EXTRA_CFLAGS="-D__phoenix__" DESTDIR="$PREFIX_WPA_SUPPLICANT_INSTALL" LIBDIR="/lib" INCDIR="/include" BINDIR="/sbin"
```
Comment (nalajcie) [inline wpa_supplicant/build.sh:46]: will this support incremental build or will it be rebuilt with every invocation?

Edit: adding `cp -a` mentioned above makes incremental build hopefully work :).

File: wpa_supplicant/build.sh [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+[ -f "${PREFIX_WPA_SUPPLICANT}/${WPA_SUPPLICANT}.tar.gz" ] || wget https://w1.fi/releases/${WPA_SUPPLICANT}.tar.gz -P "$PREFIX_WPA_SUPPLICANT" --no-check-certificate
+[ -d "$PREFIX_WPA_SUPPLICANT_SRC" ] || tar zxf "${PREFIX_WPA_SUPPLICANT}/${WPA_SUPPLICANT}.tar.gz" -C "$PREFIX_WPA_SUPPLICANT_BUILD"
+
+#
+# Apply patches
+#
+mkdir -p "$PREFIX_WPA_SUPPLICANT_MARKERS"
+
+for patchfile in "${PREFIX_WPA_SUPPLICANT}"/patches/*.patch; do
+    if [ ! -f "${PREFIX_WPA_SUPPLICANT_MARKERS}/$(basename "$patchfile").applied" ]; then
```
Comment (nalajcie) [inline wpa_supplicant/build.sh:32]: switch spaces to tabs unfortunately

File: wpa_supplicant/build.sh [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+pushd "${PREFIX_WPA_SUPPLICANT_SRC}/wpa_supplicant"
+cp "$PREFIX_WPA_SUPPLICANT/.config" .config
+make install EXTRA_CFLAGS="-D__phoenix__" DESTDIR="$PREFIX_WPA_SUPPLICANT_INSTALL" LIBDIR="/lib" INCDIR="/include" BINDIR="/sbin"
+popd
+
+mkdir -p "$PREFIX_PROG" && cp -Ra "${PREFIX_WPA_SUPPLICANT_INSTALL}/sbin/wpa_cli" "${PREFIX_PROG}/wpa_cli"
+mkdir -p "$PREFIX_PROG" && cp -Ra "${PREFIX_WPA_SUPPLICANT_INSTALL}/sbin/wpa_supplicant" "${PREFIX_PROG}/wpa_supplicant"
+mkdir -p "$PREFIX_PROG_STRIPPED" && "${CROSS}strip" -s "${PREFIX_PROG}/wpa_cli" -o "${PREFIX_PROG_STRIPPED}/wpa_cli"
+mkdir -p "$PREFIX_PROG_STRIPPED" && "${CROSS}strip" -s "${PREFIX_PROG}/wpa_supplicant" -o "${PREFIX_PROG_STRIPPED}/wpa_supplicant"
+b_install "${PREFIX_PORTS_INSTALL}/wpa_cli" /bin
```
Comment (nalajcie) [inline wpa_supplicant/build.sh:53]: maybe `/usr/bin` ?

File: build.sh (hunk 5 above)
Comment (ziemleszcz) [reply build.sh:60]: It doesn't build without additional changes in libphoenix, phoenix-rtos-kernel and phoenix-rtos-lwip (not available yet).

File: wpa_supplicant/build.sh [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+PREFIX_WPA_SUPPLICANT_MARKERS="${PREFIX_WPA_SUPPLICANT_BUILD}/markers"
+PREFIX_WPA_SUPPLICANT_INSTALL="${PREFIX_WPA_SUPPLICANT_BUILD}/install"
+
+#
+# Download and unpack
+#
+mkdir -p "$PREFIX_WPA_SUPPLICANT_BUILD"
+
+if [ -n "$CLEAN" ]; then
+	rm -rf "${PREFIX_WPA_SUPPLICANT_BUILD:?}/*"
```
Comment (kemonats) [inline wpa_supplicant/build.sh:20]:
```suggestion
	rm -rf "${PREFIX_WPA_SUPPLICANT_BUILD:?}"/*
```

File: build.sh (hunk 5 above)
Comment (nalajcie) [reply build.sh:60]: ok, changing to draft PR to avoid accidental merge by someone else :)

File: wpa_supplicant/build.sh (hunk 6 above)
Comment (ziemleszcz) [reply wpa_supplicant/build.sh:46]: Yes, works fine after changing cp options.

File: build.sh (hunk 5 above)
Comment (nalajcie) [reply build.sh:62]: is this ready for review/merge now? If so, please mark it as non-draft PR

**Review by nalajcie — APPROVED** [review]
LGTM.

**Pull request merged**